app/validators/uri_validator.rb

```diff
@@ -1,53 +1,102 @@
 # frozen_string_literal: true
 
 class UriValidator < ActiveModel::EachValidator
-  DEFAULT_ACCEPTED_SCHEMES = /^https?$/
-  DEFAULT_OPTIONAL_PARTS = %i[port].freeze
-  DEFAULT_FORBIDDEN_PARTS = %i[userinfo registry path opaque query fragment].freeze
-
-  def validate_each(record, attribute, value)
-    valid = begin
-              uri = URI.parse(value)
-              valid_scheme?(uri.scheme) && valid_host?(uri.host) && !forbidden_part?(record, uri)
-            rescue URI::InvalidURIError
-              false
-            end
-    record.errors.add(attribute, options[:message] || :invalid) unless valid
-    valid
+  def validate_each(record, attribute_name, attribute_value)
+    errors = UriPartsValidator.validate(attribute_value, self.class.options_for_record(options, record))
+    errors.each { |error| record.errors.add(attribute_name, error) }
+    errors.empty?
   end
 
-  def valid_scheme?(scheme)
-    accepted_scheme = options.fetch(:scheme, DEFAULT_ACCEPTED_SCHEMES)
-    case accepted_scheme
-    when Regexp
-      scheme =~ accepted_scheme
-    else
-      [*accepted_scheme].include? scheme
+  # This smells :reek:ManualDispatch
+  def self.options_for_record(options, record)
```

Comment:
```suggestion
  def self.options_for_record(options, record)
    return options.slice(:scheme, *UriPartsValidator::DEFAULT_PERMISSIONS_OF_PARTS.keys).transform_values { |value| value.respond_to(:call) ? value.call(record) : value.presence }
```

The `return` is only to say that we don't need the other lines below.

Reply:
Great one :+1: :clap: